whatweb

```diff
@@ -973,102 +935,93 @@ $CUSTOM_HEADERS.delete_if {|k,v| v=="" }
-				if target.is_url? and target.status.nil?
-					# assume all HTTP sites return a status
-					no_redirects=true
-					next
-				end
-
-				results = run_plugins(target)
-
-				# reporting
-				# multiple output plugins simultaneously, some stdout, some files
-				output_list.each do |o|
-					begin
-						o.out(target, target.status, results)
-					rescue => err
-						#srsly, logging failed
-						error("ERROR Logging failed: #{target.to_s} - #{err}")
-						raise if $WWDEBUG==true
-					end
-				end
-
-				# REDIRECTION
-				unless no_redirects
-					begin
-						if newtarget = target.get_redirection_target		
-							num_redirects+=1 
-							target=Target.new(newtarget)
-						else
-							no_redirects=true 
-						end
-					rescue => err
-						error("ERROR Redirection broken: #{target.to_s} - #{err}")
-						no_redirects=true
-						raise if $WWDEBUG==true
-					end
-				end
-			end # while no_redirects
-		end # Thread.new
-
-	while Thread.list.size>($MAX_THREADS+1)
-		puts "Thread list full, passing control" if $verbose>1
-		#sleep 0.5
-		Thread.pass
-	end
-end # targets.each
+# try to make a new Target object, may return nil
+def prepare_target(url)
+  begin
+    Target.new(url)
+  rescue => err
+    error(err)
+    nil
+  end
+end
 
-# close output logs
-output_list.each {|o| 
-	o.close
-}
+target_queue = Queue.new() # workers consume from this
+result_queue = Queue.new() # workers return results here for output
+result_mutex = Mutex.new()
+Thread.abort_on_exception = true if $WWDEBUG
 
-# shutdown plugins
-Plugin.registered_plugins.map {|name,plugin| plugin.shutdown }
+workers = (1..$MAX_THREADS).map do
+  Thread.new() do
+    # keep reading in root tasks until a nil is received
+    loop do
+      target = target_queue.pop
+      Thread.exit unless target
+      redirects = 0
+      # keep processing until there are no more redirects or the limit is hit
+      while target
+        begin
+          target.open
+        rescue => err
+          error("ERROR Opening: #{target} - #{err}")
+          raise if $WWDEBUG
+          next
+        end
+        result = run_plugins(target)
+        result_mutex.synchronize do
+          result_queue << [target, result]
+        end
+        redirect_url = target.get_redirection_target
+        if redirect_url
+          redirects += 1
+          target = prepare_target(redirect_url)
+          if redirects > $MAX_REDIRECTS
+            error("ERROR Too many redirects: #{target}")
+            break
+          end
+        else
+          target = nil
+        end
+      end
+    end
+  end
+end
+$targets.each do |url|
+  target = prepare_target(url)
+  target_queue << target if target
+end
+loop do
+  break if result_mutex.synchronize do
+    # more defensive than comparing against $MAX_THREADS
+    alive = workers.map { |worker| worker if worker.alive? }.compact.length
+    alive == target_queue.num_waiting and result_queue.empty?
+  end
+  begin
+    target, result = result_queue.pop(true)
```

Comment:
Using a `Queue` probably isn't essential in this case as array operations are probably atomic due to the GIL.

It would be nice if the blocking on the queue could be used.